pages/glossary/keyboard-actionable-element.md

```diff
@@ -0,0 +1,23 @@
+---
+title: Keyboard Actionable Element
+key: keyboard-actionable-element
+unambiguous: true
+objective: true
+input_aspects:
+  - Accessibility tree
+  - CSS styling
+  - DOM tree
+---
+
+An HTML element is _keyboard actionable_ if all the following are true:
+
+- the element is part of [sequential focus navigation][]; and
+- the element is [visible][] when [focused][]; and
+- it is possible to [fire a click event][] at the element by use of keyboard.
```

Comment:
Copying from https://github.com/act-rules/act-rules.github.io/pull/421/files#r562630136
> Under what circumstance would this not be true? This one is a little tricky. What do you mean with "by use of the keyboard"? If I have an on-screen cursor that I control using arrow keys, and use space to fire a click event from the cursor, does that count? Arguably it would, but I suspect that's not what you mean. Additionally, are you expecting the click event to be fired when the element is focused? Could event bubbling be involved?

Reply:
I don't think that a mouse-emulator should count. It's a separate assistive technology and I don't think that it should be taken into account.

If such an emulator would count, then we could argue that every thing satisfies 2.1.1 as soon as users have such an emulator. This is quite clearly not the intend of 2.1.1…

I think rules only test what is part of the test subject (usually a web page).

Now, if a web page implements a mouse emulation with keyboard, that is another question. I would say that such a page does satisfy 2.1.1 (maybe not in a very good way, however). In [Scrollable element is keyboard accessible](https://act-rules.github.io/rules/0ssw9k), we do have an assumption (maybe a bit weak) around this idea. (I would argue that currently a page with a mouse emulation implemented could fail the rule but satisfy 2.1.1 and it's not completely clear that this is caught by the assumption).

Anyway, here, if a page implements a mouse emulation, then yes, you can fire the click event by use of keyboard only. But that is not always the case.
```
<span tabindex="0" onclick="this.style.color='red'">Foo</span>
```
Is passing the first two conditions (it is part of sequential focus navigation, and visible when focused), but not the 3rd (that may be browser dependant, but on Chrome I do not manage to fire the click event with space or enter even when the element has focus). So I do not think that this condition is always true.

Reply:
I think we can take another approach. 
To meet the expectation of the rule, the element needs to be activated (so that a change of focus can be evaluated). So, for this rule, we don't need to check if it is possible to activate (fire a click event at) an element twice. 
I suggest we drop this definition and move being part of the sequential focus navigation and being visible when focused to the expectations of the rule. The third condition in this definition is already checked in the rule's expectation.

Reply:
The important point in this third condition is not that the element can be activated, but the fact that it can be done by using the keyboard. This is what fails Failed Example 7, and is not checked elsewhere in the rule.

Without this condition,
```html
<div role="link" onclick="location.assign('#main');" tabindex="0" id="skip-link">Skip to main content</div>
```
would pass the rule, which is a bit sad for a best practice rule.

Reply:
You are correct. I was focusing too much on the "how can we know if we can fire an event at a target" and leaving out the "using the keyboard" parte.

Still, what about changing the condition in the expectation to 
```
- when the element is [activated][] by use of keyboard, focus moves [just before][] a node of [non-repeated content after repeated content][]; and
```

Reply:
🤔 Interesting…

I'm afraid that this condition is dangerously close to triggering a universal quantification on an empty set… The almost similar
> If the element is activated by use of keyboard, focus moves…

for example, would pass on elements that are not keyboard-actionable (or we can make this even more explicit with "every time the element is activated by use of keyboard, …")

Now, clearly, the current formulation also has that problem… It can be rephrased the same way as
> Every time the element is activated, …

and would pass on a non-activatable element (whatever that means…) But the current version does have a condition that the element is actionable (as part of the keyboard actionable requirement…)

---
Also, to continue discussion… If I open the devtools (Ctrl+Shift+I, usually), go to the console, and type in some `document.getElement(foo).sendEvent(click)`, is that keyboard activation? I guess we're safe-ish because that uses something (devtools) which is not part of the content (the web page), so if that gives extra power, we don't care (the same way that we wouldn't take into account a browser extension that allows to move the pointer with the k.eyboard).